src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for finding affected ranges from over window range cache for a
+//! given set of changes (delta).
+
+use std::ops::Bound;
+
+use delta_btree_map::DeltaBTreeMap;
+use risingwave_common::row::OwnedRow;
+use risingwave_expr::window_function::{FrameBound, RowsFrameBounds};
+
+use super::over_partition::CacheKey;
+
+// -------------------------- ↓ PUBLIC INTERFACE ↓ --------------------------
+
+/// Merge several `ROWS` frames into one super frame. The returned super frame is
+/// ensured to not be a pure preceding/following frame, which means that `CURRENT ROW`
+/// is always included in the returned frame.
+pub(super) fn merge_rows_frames(rows_frames: &[&RowsFrameBounds]) -> RowsFrameBounds {
+    if rows_frames.is_empty() {
+        // this can simplify our implementation
+        return RowsFrameBounds {
+            start: FrameBound::CurrentRow,
+            end: FrameBound::CurrentRow,
+        };
+    }
+
+    let none_as_max_cmp = |x: &Option<usize>, y: &Option<usize>| match (x, y) {
+        // None should be maximum (unbounded)
+        (None, None) => std::cmp::Ordering::Equal,
+        (None, Some(_)) => std::cmp::Ordering::Greater,
+        (Some(_), None) => std::cmp::Ordering::Less,
+        (Some(x), Some(y)) => x.cmp(y),
+    };
+
+    // Note that the following two both use `max_by`, unlike when handling `RANGE` frames,
+    // because here for `ROWS` frames, offsets are unsigned. We convert all pure preceding/
+    // following `ROWS` frames into one containing the `CURRENT ROW`.
+    let start = rows_frames
+        .iter()
+        .map(|bounds| bounds.start.n_preceding_rows())
+        .max_by(none_as_max_cmp)
+        .unwrap();
+    let end = rows_frames
+        .iter()
+        .map(|bounds| bounds.end.n_following_rows())
+        .max_by(none_as_max_cmp)
+        .unwrap();
```

Comment:
Why both `start` and `end` use `max_by`?

I thought `start` will be `min_by`? Perhaps I don't understand the semantics of `merge_rows_frames` sufficiently.

What's the merge constraint?

Reply:
Because `n_preceding_rows` and `n_following_rows` return `usize`. If the `start` of a frame is `3 FOLLOWING`, `start.n_preceding_rows()` will return `0`.

This function basically behaves as if there's an additional hidden frame `ROWS BETWEEN CURRENT ROW AND CURRENT ROW`.